File: app/src/main/java/io/github/fvasco/pinpoi/importer/GeoRssImporter.java
package io.github.fvasco.pinpoi.importer;

/**
 * Import simple geo RSS/Atom
 *
 * @author Francesco Vasco
 */
public class GeoRssImporter extends AbstractXmlImporter {
    @Override
    protected void handleStartTag() {
        switch (tag) {
            case "item": // RSS
            case "entry": // ATOM
                newPlacemark();
                break;
        }
    }

    @Override
    protected void handleEndTag() {
        if (placemark == null) return;
        switch (tag) {
            case "item": // RSS
            case "entry": // ATOM
                importPlacemark();
                break;
            case "title":
                placemark.setName(text);
                break;
            case "description": // Atom
            case "summary": // RSS
                placemark.setDescription(text);
                break;
            case "point": // Atom
                String[] parts = text.split("[ ,]+", 3);
                if (parts.length >= 2) {
                    placemark.setLatitude(Float.parseFloat(parts[0]));
                    placemark.setLongitude(Float.parseFloat(parts[1]));
                }
                break;
            case "lat": // RSS
                placemark.setLatitude(Float.parseFloat(text));
                break;
            case "long": // RSS
                placemark.setLongitude(Float.parseFloat(text));
                break;
        }
    }
}

File: app/src/main/kotlin/io/github/fvasco/pinpoi/importer/GeoRssImporter.kt
package io.github.fvasco.pinpoi.importer

import io.github.fvasco.pinpoi.util.Coordinates

/**
 * Import simple geo RSS/Atom

 * @author Francesco Vasco
 */
class GeoRssImporter : AbstractXmlImporter() {
    override fun handleStartTag() {
        when (tag) {
            "item" // RSS
                , "entry" // ATOM
            -> newPlacemark()
        }
    }

    override fun handleEndTag() {
        val p = placemark ?: return
        when (tag) {
            "item" // RSS
                , "entry" // ATOM
            -> importPlacemark()

            "title" -> p.name = text

            "description" // Atom
                , "summary" // RSS
            -> p.description = text

            "point" // Atom
            -> {
                val parts = text.split("[ ,]+".toRegex(), 3).toTypedArray()
                if (parts.size >= 2) {
                    p.coordinates = Coordinates(java.lang.Float.parseFloat(parts[0]), java.lang.Float.parseFloat(parts[1]))
                }
            }

        // RSS
            "lat" -> p.coordinates = p.coordinates.copy(latitude = java.lang.Float.parseFloat(text))
            "long" -> p.coordinates = p.coordinates.copy(longitude = java.lang.Float.parseFloat(text))
        }
    }
}
